packages/core/echo/echo-schema/src/effect/echo-handler.ts

```diff
@@ -405,44 +496,48 @@ export class EchoReactiveHandlerImpl extends EchoReactiveHandler implements Reac
     return deletedElements;
   }
 
-  arraySort(target: any, path: KeyPath, compareFn?: (v1: any, v2: any) => number): any[] {
+  arraySort(target: ProxyTarget, path: KeyPath, compareFn?: (v1: any, v2: any) => number): any[] {
     const fullPath = this._getPropertyMountPath(target, path);
 
-    this._objectCore.change((doc) => {
+    target[symbolInternals].core.change((doc) => {
       const array = getDeep(doc, fullPath);
       invariant(Array.isArray(array));
       const sortedArray = [...array].sort(compareFn);
       assignDeep(doc, fullPath, sortedArray);
     });
 
-    return target;
+    return target as EchoArrayTwoPointO<any>;
   }
 
-  arrayReverse(target: any, path: KeyPath): any[] {
+  arrayReverse(target: ProxyTarget, path: KeyPath): any[] {
     const fullPath = this._getPropertyMountPath(target, path);
 
-    this._objectCore.change((doc) => {
+    target[symbolInternals].core.change((doc) => {
       const array = getDeep(doc, fullPath);
       invariant(Array.isArray(array));
       const reversedArray = [...array].reverse();
       assignDeep(doc, fullPath, reversedArray);
     });
 
-    return target;
+    return target as EchoArrayTwoPointO<any>;
   }
 
-  getMeta(): ObjectMeta {
-    const target: any = { [symbolPath]: [], [symbolNamespace]: META_NAMESPACE };
-    return createReactiveProxy(target, this) as ObjectMeta;
+  getMeta(target: ProxyTarget): ObjectMeta {
+    const metaTarget: ProxyTarget = {
+      [symbolInternals]: target[symbolInternals],
```

Comment:
dataPath doesn't include namespace, so should be a separate map for meta proxy or full path should be used when doing lookup

Reply:
for targets map?

yeah, its also broken for expandos - see my comment on where it is declared

Reply:
I mean imagine you have `struct({ keys: S.array(S.any) })`
here you will create an object and cache it using `['keys']` key
https://github.com/dxos/dxos/blob/2ec80d874299475b4718b958e7447f503c9ec3d2/packages/core/echo/echo-schema/src/effect/echo-handler.ts#L249
but then you request a meta object and when someone accesses keys they get the cache target from `data` namespace

Reply:
done